File: export_data_scripts/fcs_script.m
file_path = '<FILE-PATH>';

% Channel custom names
channel_names_json = '<CHANNEL-NAMES-JSON>';
channel_names = jsondecode(channel_names_json);

% Open the file
metadata = struct('enabled_channels', [], 'correlations', [], 'num_acquisitions', [], 'acquisition_time', [], 'bin_width', [], 'notes', []);

fid = fopen(file_path, 'rb');

if fid == -1
    error('Unable to open the file');
end


% First 4 bytes must be FCS1
first_bytes = fread(fid, 4, 'char')';

if ~isequal(first_bytes, 'FCS1')
    fprintf('Invalid data file');
    fclose(fid);
    return;
end

% Read metadata json length
json_metadata_length = fread(fid, 1, 'uint32');

% Read metadata from file
json_metadata_data = fread(fid, json_metadata_length, 'char')';
json_metadata_str = char(json_metadata_data);
metadata = jsondecode(json_metadata_str);

% Read g2_correlations JSON length
g2_correlations_json_length = fread(fid, 1, 'uint32');

% Extract lag_index and g2_correlations from JSON
g2_correlations_json_data = fread(fid, g2_correlations_json_length, 'char')';
g2_correlations_json_str = char(g2_correlations_json_data);
g2_correlations_json = jsondecode(g2_correlations_json_str);

lag_index = g2_correlations_json.lag_index;
g2_correlations = g2_correlations_json.g2_correlations;

% Function to get channel name
function name = get_channel_name(channel_id, channel_names, truncate_len)
    if nargin < 3
        truncate_len = 5;
    end
    field_name = ['x' num2str(channel_id)];
    if isfield(channel_names, field_name)
        custom_name = channel_names.(field_name);
        if length(custom_name) > truncate_len
            custom_name = [custom_name(1:truncate_len) '...'];
        end
        name = [custom_name ' (Ch' num2str(channel_id + 1) ')'];
    else
        name = ['Channel ' num2str(channel_id + 1)];
    end
end

% Enabled channels
if ~isempty(metadata.enabled_channels)
    channel_names_str = arrayfun(@(ch) get_channel_name(ch, channel_names, 5), metadata.enabled_channels, 'UniformOutput', false);
    disp(['Enabled channels: ' strjoin(channel_names_str, ', ')]);
end

% Correlations (info about correlated channels)
if ~isempty(metadata.correlations)
    correlated_channels = metadata.correlations + 1;
    disp(['Channels correlations: ']);
    disp(mat2str(correlated_channels));
end

% Number of acquisitions
if ~isempty(metadata.num_acquisitions)
    disp(['Number of acquisitions: ' num2str(metadata.num_acquisitions)]);
end

% Acquisition time (duration of the acquisition)
if ~isempty(metadata.acquisition_time)
    num_acquisitions = metadata.num_acquisitions;
    if ~isempty(num_acquisitions)
        disp(['Acquisition time: ' num2str(sprintf('%.2f', metadata.acquisition_time / 1000 / num_acquisitions)) 's']);
    end
end

% Bin width
if ~isempty(metadata.bin_width)
    disp(['Bin width: ' num2str(metadata.bin_width) ' us']);
end

% Tau
disp(['Tau (lag index): ' mat2str(lag_index)]);

% Notes
if ~isempty(metadata.notes)
    disp(['Notes: ' num2str(metadata.notes)]);
end


% Parse correlations data
g2_with_mean = cell(size(g2_correlations, 1), 1);
for idx = 1:size(g2_correlations, 1)
    corr = g2_correlations{idx};
    correlations_mean = corr{2}(1, :);
    g2_vectors = corr{2}(2:end, :); 
    g2_vector_cells = num2cell(g2_vectors, 2);
    g2_with_mean{idx} = {corr{1}, correlations_mean, g2_vector_cells}; 
end

% Plot G(t) correlations
num_plots = numel(g2_with_mean);
num_plots_per_row = 1;
if num_plots < 2
    num_plots_per_row = 1;
elseif num_plots > 1 && num_plots < 4
    num_plots_per_row = 2;
elseif num_plots >= 4
    num_plots_per_row = 4;
end
if num_plots >= 12
    num_plots_per_row = 6;
end

num_rows = ceil(num_plots / num_plots_per_row);

fig = figure('Name', 'Fluorescence Spectroscopy Correlations');
lag_index_plot = lag_index + 1e-6;

for i = 1:num_plots
    corr = g2_with_mean{i};
    channel_info = corr{1};
    channel1 = channel_info(1);
    channel2 = channel_info(2);
    correlations_mean = corr{2};  
    single_gt_correlations = corr{3};  

    ax = subplot(num_rows, num_plots_per_row, i);
    
    % Plot correlations mean
    plot(lag_index_plot, correlations_mean, 'DisplayName', 'G(\tau) mean');
    hold on;

    % Plot single gt correlations vectors
    for data_index = 1:numel(single_gt_correlations)
        current_data = single_gt_correlations{data_index};
        plot(lag_index_plot, current_data, 'DisplayName', ['G(\tau) ' num2str(data_index)]);
    end
   
    ch1_name = get_channel_name(channel1, channel_names, 5);
    ch2_name = get_channel_name(channel2, channel_names, 5);
    
    hold off;
    set(gca, 'XScale', 'log');
    xlim([1, max(lag_index_plot)]);
    xlabel('\tau (\mus)');
    ylabel('G(\tau)');
    title([ch1_name ' - ' ch2_name]);
    grid on;
    legend('Location', 'northeast');
end

hold off;

File: export_data_scripts/fcs_script.py
import struct
from matplotlib.gridspec import GridSpec
import matplotlib.pyplot as plt


file_path = "<FILE-PATH>"
print("Using data file: " + file_path)

# Read bin data
with open(file_path, "rb") as f:
    # first 4 bytes must be FCS1
    # 'FCS1' is an identifier for fcs bin files
    if f.read(4) != b"FCS1":
        print("Invalid data file")
        exit(0)

    # read metadata from file
    (json_length,) = struct.unpack("I", f.read(4))
    null = None
    metadata = eval(f.read(json_length).decode("utf-8"))

    # Read g2_correlations JSON length
    (g2_correlations_json_length,) = struct.unpack("I", f.read(4))
    g2_correlations_json_string = f.read(g2_correlations_json_length).decode("utf-8")
    g2_correlations_json = eval(g2_correlations_json_string)

    # Extract lag_index and g2_correlations from JSON
    lag_index = g2_correlations_json["lag_index"]
    g2_correlations = g2_correlations_json["g2_correlations"]

    # ENABLED CHANNELS
    if "enabled_channels" in metadata and metadata["enabled_channels"]:
        print(
            "Enabled channels: "
            + (
                ", ".join(
                    ["Channel " + str(ch + 1) for ch in metadata["enabled_channels"]]
                )
            )
        )

    # CORRELATIONS (info about correlated channels)
    if "correlations" in metadata and metadata["correlations"] is not None:
        correlated_channels = [
            [channel + 1 for channel in pair] for pair in metadata["correlations"]
        ]
        print("Channels correlations: " + str(correlated_channels))

    # NUMBER OF ACQUISITIONS
    if "num_acquisitions" in metadata and metadata["num_acquisitions"] is not None:
        print("Number of acquisitions: " + str(metadata["num_acquisitions"]))

    # ACQUISITION TIME (duration of the acquisition)
    if "acquisition_time" in metadata and metadata["acquisition_time"] is not None:
        num_acquisition = (
            int(metadata["num_acquisition"])
            if "num_acquisition" in metadata and metadata["num_acquisition"] is not None
            else 1
        )
        print(
            "Acquisition time: "
            + "{:.2f}".format(
                round(metadata["acquisition_time"] / 1000 / num_acquisition, 2)
            )
            + "s"
        )

    # BIN WIDTH
    if "bin_width" in metadata and metadata["bin_width"] is not None:
        print("Bin width: " + str(metadata["bin_width"]) + "\u00B5s")

    # TAU
    print("Tau (lag index):", lag_index)

    # NOTES
    if "notes" in metadata and metadata["notes"] is not None:
        print("Notes: " + str(metadata["notes"]))

    # Plot G(t) correlations
    num_plots = len(g2_correlations)
    num_plots_per_row = 1
    if num_plots < 2:
        num_plots_per_row = 1
    if num_plots > 1 and num_plots < 4:
        num_plots_per_row = 2
    if num_plots >= 4:
        num_plots_per_row = 4
    if num_plots >= 12:
        num_plots_per_row = 6

    num_rows = (num_plots + num_plots_per_row - 1) // num_plots_per_row

    fig = plt.figure(figsize=(12, 3 * num_rows))
    fig.suptitle("Fluorescence Spectroscopy Correlations")
    gs = GridSpec(num_rows, num_plots_per_row, figure=fig)

    for i, ((channel1, channel2), data_list) in enumerate(g2_correlations):
        row = i // num_plots_per_row
        col = i % num_plots_per_row
        ax = fig.add_subplot(gs[row, col])
        for data_index, data in enumerate(data_list):
            label = f"G(τ) {data_index + 1}" if data_index != 0 else f"G(τ) mean"
            ax.plot(lag_index, data, label=label)
        ax.set_xlabel("τ(μs)")
        ax.set_ylabel("G(τ)")
        ax.set_title(f"Channel {channel1 + 1}- Channel {channel2 + 1}")
        ax.set_xscale("log")
        ax.grid(True)
        ax.legend()

    plt.tight_layout()
    plt.show()
